### PR TITLE
Add FinalOrAbstractClassLinter with tests

### PR DESCRIPTION
**Pull request by icechen1**
Open sourcing a linter to enforce `final` or `abstract` qualifiers on all class declarations (except on interfaces), also includes some test fixtures.

This is a followup to #87.

### CODE REVIEW

File: src/Linters/FinalOrAbstractClassLinter.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+};
+
+use type Facebook\HHAST\Linters\{ASTLinter, ASTLintError};
+
+#
+# This linter ensures we always qualify classes as final or abstract
+#
+final class FinalOrAbstractClassLinter extends ASTLinter<ClassishDeclaration> {
+
+	<<__Override>>
```
Comment (fredemmott) [inline src/Linters/FinalOrAbstractClassLinter.php]: hackfmt :)

- no newline above
- two space indents

File: src/Linters/FinalOrAbstractClassLinter.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  ClassishDeclaration,
+  EditableNode,
+  FinalToken,
+  AbstractToken,
+  ClassToken,
+};
+
+use type Facebook\HHAST\Linters\{ASTLinter, ASTLintError};
+
+#
```
Comment (fredemmott) [inline src/Linters/FinalOrAbstractClassLinter.php]: `/**` doccomments

File: src/Linters/FinalOrAbstractClassLinter.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	protected static function getTargetType(): classname<ClassishDeclaration> {
+		return ClassishDeclaration::class;
+	}
+
+	<<__Override>>
+	public function getLintErrorForNode(
+		ClassishDeclaration $node,
+		vec<EditableNode> $_parents,
+		): ?ASTLintError<ClassishDeclaration> {
+		# ensure we are looking at a class declaration
```
Comment (fredemmott) [inline src/Linters/FinalOrAbstractClassLinter.php]: // comments.

File: src/Linters/FinalOrAbstractClassLinter.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+			return null;
+		}
+		
+		# check if the ClassishDeclaration has modifiers
+		$modifiers = $node->getModifiers();
+		$found = false;
+		if ($modifiers !== null) {
+			foreach ($modifiers->traverse() as $mod) {
+				if ($mod is FinalToken || $mod is AbstractToken) {
+					$found = true;
```
Comment (fredemmott) [inline src/Linters/FinalOrAbstractClassLinter.php]: I think you can just do `return null` here, and return the error without an if condition below

File: tests/FinalOrAbstractClassLinterTest.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  use LinterTestTrait;
+
+  protected function getLinter(
+    string $file,
+  ): Linters\BaseLinter {
+    return Linters\FinalOrAbstractClassLinter::fromPath($file);
+  }
+
+  public function getCleanExamples(): array<array<string>> {
+    return [['<?hh final class test {}'], ['<?hh abstract class test {}'], ['<?hh interface test {}']];
```
Comment (fredemmott) [inline tests/FinalOrAbstractClassLinterTest.php]: - this should be multiple lines
- existing tests are using PHP arrays, but now we're on HackTest, vecs should be usable instead

File: tests/fixtures/FinalOrAbstractClassLinter/classes.php.in [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+}
+
+abstract class test4 {
+
+}
+
+interface interface1 {
+
+}
```
Comment (fredemmott) [inline tests/fixtures/FinalOrAbstractClassLinter/classes.php.in]: newline at EOF

Comment (fredemmott) [reply tests/fixtures/FinalOrAbstractClassLinter/classes.php.in]: also:
- add an abstract final class
- add XHP classes
- add traits

Comment (icechen1) [reply tests/fixtures/FinalOrAbstractClassLinter/classes.php.in]: Thanks for the feedback above! I've fixed them just now.

I wonder if we need to handle the `abstract final` class for this linter. That feels like a syntax error?

File: tests/FinalOrAbstractClassLinterTest.php (hunk 5 above)
Comment (icechen1) [reply tests/FinalOrAbstractClassLinterTest.php]: Hmm, not sure how we can use `vec` here? `LinterTestTrait` still requires `array<array<string>>` to be returned.

File: src/Linters/FinalOrAbstractClassLinter.php (hunk 4 above)
Comment (icechen1) [reply src/Linters/FinalOrAbstractClassLinter.php]: Right on, thanks!

File: tests/fixtures/FinalOrAbstractClassLinter/classes.php.in (hunk 6 above)
Comment (fredemmott) [reply tests/fixtures/FinalOrAbstractClassLinter/classes.php.in]: `abstract final` is allowed in Hack; this is useful for classes that only have static methods (especially if you want to reference them via `classname<SomeSuperclass>`)

File: src/Linters/FinalOrAbstractClassLinter.php [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    // check if the ClassishDeclaration has modifiers
+    $modifiers = $node->getModifiers();
+    $found = false;
+    if ($modifiers !== null) {
+      foreach ($modifiers->traverse() as $mod) {
+        $isFinal = $mod is FinalToken;
+        $isAbstract = $mod is AbstractToken;
+        if (($isFinal || $isAbstract) && !($isFinal && $isAbstract)) {
```
Comment (fredemmott) [inline src/Linters/FinalOrAbstractClassLinter.php]: - a specific modifier can not be both Final and Abstract - this check doesn't make sense; perhaps you want `$is_final = $is_final || $mod is FinalToken`, and move the if outside the loop?
- final and abstract is fine, the condition after the && isn't needed.

File: tests/FinalOrAbstractClassLinterTest.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    string $file,
+  ): Linters\BaseLinter {
+    return Linters\FinalOrAbstractClassLinter::fromPath($file);
+  }
+
+  public function getCleanExamples(): vec<array<string>> {
+    return vec[
+      ['<?hh final class test {}'],
+      ['<?hh abstract class test {}'],
+      ['<?hh interface test {}'],
```
Comment (fredemmott) [inline tests/FinalOrAbstractClassLinterTest.php]: abstract final
trait
xhp

File: src/Linters/FinalOrAbstractClassLinter.php (hunk 7 above)
Comment (fredemmott) [reply src/Linters/FinalOrAbstractClassLinter.php]: whoops, given the combination of the two, you can do:

```
foreach ($foo as $mod) {
  if ($mod is FinalToken || $mod is AbstractToken) {
    return null;
  }
}
```

File: tests/fixtures/FinalOrAbstractClassLinter/classes.php.in (hunk 6 above)
Comment (icechen1) [reply tests/fixtures/FinalOrAbstractClassLinter/classes.php.in]: Gotcha, the only reference I can find about this is a rejected PHP proposal for `abstract final` classes. So I was not sure if the same applies to Hack. Good to know!

File: src/Linters/FinalOrAbstractClassLinter.php (hunk 7 above)
Comment (icechen1) [reply src/Linters/FinalOrAbstractClassLinter.php]: Oops I was trying out something and forgot to remove it. Thanks for catching it

File: tests/FinalOrAbstractClassLinterTest.php (hunk 8 above)
Comment (icechen1) [reply tests/FinalOrAbstractClassLinterTest.php]: Done!

File: src/Linters/FinalOrAbstractClassLinter.php [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        }
+      }
+    }
+    return new ASTLintError(
+      $this,
+      'Class should always be declared abstract or final',
+      $node,
+    );
+  }
+}
```
Comment (fredemmott) [inline src/Linters/FinalOrAbstractClassLinter.php]: newline at eof

**Comment by fredemmott** [comment]
fyi, running `bin/hhast-lint` picks up the newline at EOF issues

**Comment by icechen1** [comment]
Thanks, not sure why the VSCode Hack extension isn't picking up on these, maybe I goofed the config somewhere

**Comment by fredemmott** [comment]
Thanks!

**Pull request merged**